test/sql-luatest/datetime_test.lua

```diff
@@ -1640,3 +1640,632 @@ g.test_datetime_29_2 = function()
+
+        rows = box.execute(sql, {dt.new({sec = 47, nsec = 176318368})}).rows
+        res = {{176}}
+        t.assert_equals(rows, res)
+
+        rows = box.execute(sql, {dt.new({sec = 59, nsec = 29})}).rows
+        res = {{0}}
+        t.assert_equals(rows, res)
+    end)
+end
+
+g.test_datetime_30_15 = function()
+    g.server:exec(function()
+        local t = require('luatest')
+        local dt = require('datetime')
+        local dt0 = dt.new({year = 2000, month = 4, day = 5, hour = 6,
+                            min = 33, sec = 22, nsec = 523999111})
+
+        local sql = [[SELECT date_part('microseconds', ?);]]
+        local rows = box.execute(sql, {dt0}).rows
+        local res = {{523999}}
+        t.assert_equals(rows, res)
+
+        rows = box.execute(sql, {dt.new({sec = 1, nsec = 13123546})}).rows
+        res = {{13123}}
+        t.assert_equals(rows, res)
+
+        rows = box.execute(sql, {dt.new({sec = 20, nsec = 591256})}).rows
+        res = {{591}}
+        t.assert_equals(rows, res)
+
+        rows = box.execute(sql, {dt.new({sec = 47, nsec = 176318368})}).rows
+        res = {{176318}}
+        t.assert_equals(rows, res)
+
+        rows = box.execute(sql, {dt.new({sec = 59, nsec = 29})}).rows
+        res = {{0}}
+        t.assert_equals(rows, res)
+    end)
+end
+
+g.test_datetime_30_16 = function()
+    g.server:exec(function()
+        local t = require('luatest')
+        local dt = require('datetime')
+        local dt0 = dt.new({year = 2000, month = 4, day = 5, hour = 6,
+                            min = 33, sec = 22, nsec = 523999111})
+
+        local sql = [[SELECT date_part('nanoseconds', ?);]]
+        local rows = box.execute(sql, {dt0}).rows
+        local res = {{523999111}}
+        t.assert_equals(rows, res)
+
+        rows = box.execute(sql, {dt.new({sec = 1, nsec = 13123546})}).rows
+        res = {{13123546}}
+        t.assert_equals(rows, res)
+
+        rows = box.execute(sql, {dt.new({sec = 20, nsec = 591256})}).rows
+        res = {{591256}}
+        t.assert_equals(rows, res)
+
+        rows = box.execute(sql, {dt.new({sec = 47, nsec = 176318368})}).rows
+        res = {{176318368}}
+        t.assert_equals(rows, res)
+
+        rows = box.execute(sql, {dt.new({sec = 59, nsec = 29})}).rows
+        res = {{29}}
+        t.assert_equals(rows, res)
+    end)
+end
+
+g.test_datetime_30_17 = function()
+    g.server:exec(function()
+        local t = require('luatest')
+        local dt = require('datetime')
+        local dt0 = dt.new({year = 2000, month = 4, day = 5, hour = 6,
+                            min = 33, sec = 22, nsec = 523999111})
+
+        local sql = [[SELECT date_part('epoch', ?);]]
+        local rows = box.execute(sql, {dt0}).rows
+        local res = {{954916402}}
+        t.assert_equals(rows, res)
+
+        dt0 = dt.new({year = 1, month = 1, day = 15, hour = 6, min = 12})
+        rows = box.execute(sql, {dt0}).rows
+        res = {{-62134364880}}
```

Comment:
Why is the result negative while the datetime object has no negative fields?

Reply:
The epoch began on 1970-01-01 00:00:00. Since 0001-01-15 06:12:00 happened earlier, value of epoch is negative.

Reply:
Oh, right, I forgot, thanks for the explanation.